web/api/backend.go

```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	jwtmiddleware "github.com/auth0/go-jwt-middleware"
+	"github.com/dgrijalva/jwt-go"
+
+	"github.com/gorilla/mux"
+)
+
+// Claims object that will be encoded to a JWT.
+// We add jwt.StandardClaims as an embedded type, to provide fields like expiry time
+type Claims struct {
+	UploadCredentials map[string]interface{}
+	jwt.StandardClaims
+}
+
+var jwtKey = []byte(os.Getenv("SECRET_KEY"))
+var creds WrapperCommand
+
+// TokenHandler to create the Token
+func TokenHandler(w http.ResponseWriter, r *http.Request) {
+	setupResponse(&w, r)
+	if (*&r).Method == "OPTIONS" {
+		return
+	}
+
+	// Get the JSON body and decode into credentials
+	err := json.NewDecoder(r.Body).Decode(&creds)
+	if err != nil {
+		// If the structure of the body is wrong, return an HTTP error
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if len(creds.Credential) == 0 {
+		w.WriteHeader(http.StatusBadRequest)
+		io.WriteString(w, `{"error":"cannot make token without credentials"}`)
+		return
+	}
+
+	// Declare the expiration time of the token
+	// here, we have kept it as 5 minutes
+	expirationTime := time.Now().Add(1440 * time.Minute)
+	// Create the JWT claims, which includes the username and expiry time
+	claims := &Claims{
+		UploadCredentials: creds.Credential,
+
+		StandardClaims: jwt.StandardClaims{
+			// In JWT, the expiry time is expressed as unix milliseconds
+			ExpiresAt: expirationTime.Unix(),
+		},
+	}
+
+	// Declare the token with the algorithm used for signing, and the claims
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	// Create the JWT string
+	tokenString, err := token.SignedString(jwtKey)
+	if err != nil {
+		// If there is an error in creating the JWT return an internal server error
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	io.WriteString(w, `{"token":"`+tokenString+`"}`)
+
+}
+
+// AuthHandler checks if token is valid. Returning a 401 status to the client if it is not valid.
+func AuthHandler(next http.Handler) http.Handler {
+	jwtMiddleware := jwtmiddleware.New(jwtmiddleware.Options{
+		ValidationKeyGetter: func(token *jwt.Token) (interface{}, error) {
+			return jwtKey, nil
+		},
+		SigningMethod: jwt.SigningMethodHS256,
+	})
+	return jwtMiddleware.Handler(next)
+}
+
+//NoTokenHandler for the case when a cached token is not used
```

Comment:
"NoCredentialsHandler for the case when user does not upload credentials?" How does "cacheCreds" work?

Reply:
Fixed! And cacheCreds is so that when the JSON body is requested with the parts of the command, it will not run the risk of tampering with the global creds variable that is used when the json file is uploaded